### PR TITLE
SAA-1801: cancelled activity attendance recording

### PR DESCRIPTION
**Pull request by dafhoustonmoj**
Creating attendance records for cancelled activity use cancelled time for the attendance recorded time. Need to allow attendance records to be edited for uncancelling an activity on the day

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt [hunk 1]
```diff
@@ -227,7 +228,10 @@ data class Attendance(
                   (
                     this.attendanceReason?.attended == false && this.issuePayment == false
                     ) ||
-                    (this.recordedTime!!.isAfter(LocalDate.now().atStartOfDay()) || this.recordedTime!!.isEqual(LocalDate.now().atStartOfDay()))
+                    (this.recordedTime!!.isAfter(LocalDate.now().atStartOfDay()) || this.recordedTime!!.isEqual(LocalDate.now().atStartOfDay())) ||
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:231]:
```
      this.scheduledInstance.sessionDate.isAfter(LocalDate.now().minusDays(7)) && (
          this.status() == AttendanceStatus.WAITING || (
              this.status == AttendanceStatus.COMPLETED && (
                  this.attendanceReason?.attended == false && this.issuePayment == false || 
                    this.recordedTime!!.toLocalDate().isEqual(LocalDate.now()) ||
                    // attendance records are created on the day.
                    // For a cancelled attendance records we need to consider the session date, not the recorded time to allow uncancelling on the day
                    (this.scheduledInstance.sessionDate.isEqual(LocalDate.now()) && this.attendanceReason?.code == AttendanceReasonEnum.CANCELLED)
                  )
              )
          )
      )
   ```

Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:231]: easier on the brain. Will update it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt [hunk 2]
```diff
@@ -211,26 +211,25 @@ data class Attendance(
 
   /*
    Attendance is editable if:
-   1. If no attendance has been recorded AND it is within 14 days of the activity session date.
+   1. If no attendance has been recorded AND it is within 7 days of the activity session date.
    2. If attendance is Not Attended & Unpaid AND it is within 7 days of the activity session date.
    3. If attendance is Attended AND it is the same day as the attendance was set as Attended.
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:216]: "Same day" means "today"?

Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:216]: Looking at this code:

```kotlin
this.attendanceReason?.attended == false && this.issuePayment == false ||
```

If `this.attendanceReason?.attended == true && this.issuePayment == false` then what should happen?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt [hunk 3]
```diff
@@ -211,26 +211,25 @@ data class Attendance(
...
-                  )
-              )
+      this.scheduledInstance.sessionDate.isAfter(LocalDate.now().minusDays(7)) && (
+        this.status() == AttendanceStatus.WAITING || (
+          this.status == AttendanceStatus.COMPLETED && (
+            this.attendanceReason?.attended == false && this.issuePayment == false ||
+              this.recordedTime!!.toLocalDate().isEqual(LocalDate.now()) ||
+              // attendance records are created on the day.
+              // For a cancelled attendance records we need to consider the session date, not the recorded time to allow uncancelling on the day
+              (this.scheduledInstance.sessionDate.isEqual(LocalDate.now()) && this.attendanceReason?.code == AttendanceReasonEnum.CANCELLED)
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:229]: To be sure you can only edit the attendance if it was cancelled the instance was scheduled for today?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt (hunk 2 above)
Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:216]: Yes, in that if you update an attendance you have the rest of the day to change it before it's locked

Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Attendance.kt:216]: you can still update it if recording it today

**Review by munslowl — APPROVED** [review]

**Pull request merged**